CrazyCanvas/Source/ECS/Systems/Player/WeaponSystem.cpp

```diff
@@ -96,25 +118,66 @@ void WeaponSystem::Tick(LambdaEngine::Timestamp deltaTime)
 			continue;
 		}
 
-		const int onCooldown = weaponComponent.CurrentCooldown > 0.0f;
-		weaponComponent.CurrentCooldown -= onCooldown * dt;
+		const bool hasAmmo		= weaponComponent.CurrentAmmunition > 0;
+		const bool isReloading	= weaponComponent.ReloadClock > 0.0f;
+		if (!hasAmmo && !isReloading)
+		{
+			StartReload(weaponComponent);
+		}
+
+		const bool onCooldown = weaponComponent.CurrentCooldown > 0.0f;
+		if (onCooldown)
+		{
+			weaponComponent.CurrentCooldown -= dt;
+		}
+
+		if (isReloading)
+		{
+			LOG_INFO("Reloading");
+
+			weaponComponent.ReloadClock -= dt;
+			if (weaponComponent.ReloadClock < 0.0f)
+			{
+				weaponComponent.ReloadClock			= 0.0f;
+				weaponComponent.CurrentAmmunition	= 5;
 
-		if (Input::GetMouseState().IsButtonPressed(EMouseButton::MOUSE_BUTTON_FORWARD) && !onCooldown)
+				LOG_INFO("Reload Finish");
+			}
+		}
+
+		// Reload if we are not reloading
+		if (Input::IsKeyDown(EKey::KEY_R) && !isReloading)
+		{
+			StartReload(weaponComponent);
+		}
+		else if (!onCooldown) // If we did not hit the reload try and shoot
 		{
-			const PositionComponent& positionComp = pPositionComponents->GetConstData(playerEntity);
-			const RotationComponent& rotationComp = pRotationComponents->GetConstData(playerEntity);
-			const VelocityComponent& velocityComp = pVelocityComponents->GetConstData(playerEntity);
+			if (Input::GetMouseState().IsButtonPressed(EMouseButton::MOUSE_BUTTON_FORWARD))
+			{
+				const PositionComponent& positionComp = pPositionComponents->GetConstData(playerEntity);
+				const VelocityComponent& velocityComp = pVelocityComponents->GetConstData(playerEntity);
+				const RotationComponent& rotationComp = pRotationComponents->GetConstData(playerEntity);
+
+				TryFire(EAmmoType::AMMO_TYPE_PAINT, weaponComponent, positionComp.Position + glm::vec3(0.0f, 1.0f, 0.0f), rotationComp.Quaternion, velocityComp.Velocity);
+			}
+			else if (Input::GetMouseState().IsButtonPressed(EMouseButton::MOUSE_BUTTON_BACK))
+			{
+				const PositionComponent& positionComp = pPositionComponents->GetConstData(playerEntity);
+				const VelocityComponent& velocityComp = pVelocityComponents->GetConstData(playerEntity);
+				const RotationComponent& rotationComp = pRotationComponents->GetConstData(playerEntity);
 
-			Fire(weaponComponent, positionComp.Position, rotationComp.Quaternion, velocityComp.Velocity);
+				TryFire(EAmmoType::AMMO_TYPE_WATER, weaponComponent, positionComp.Position + glm::vec3(0.0f, 1.0f, 0.0f), rotationComp.Quaternion, velocityComp.Velocity);
+			}
```

Comment:
This means you can't interrupt a reload, i.e. change to firing water if I'm in the middle of reloading paint. Is that an expected behaviour? 
also, does water and paint share ammunition at the moment?

Reply:
* You can interrupt a reload, however, I have not tested if you can reload and interrupt with water, but it should work. What this does is so that you cannot hit reload and fire in the same frame. I can change this if we would like to have this behavior.
* Yes, water and paint share ammunition. This is due to the fact that we have not really decided how this should work. We should discuss this and decide.

Reply:
Alright, Thank you for clarifying! =)